### PR TITLE
Unify and simplify the task-creation builtins

### PR DESCRIPTION
**Pull request by rjmccall**
- Only allow one of these builtins (`builtin "createAsyncTask"`) to survive into SIL
- Remove the explicit metadata argument and recreate it from the substitutions in IRGen
- Check a bunch of correctness conditions on the builtin and tighten up the associated tests
- Add the ability for builtins to have labeled and defaulted arguments
- Add new builtins that use the labeled and defaulted arguments

### CODE REVIEW

File: include/swift/AST/Builtins.def [hunk 1]
```diff
@@ -968,48 +952,68 @@ BUILTIN_MISC_OPERATION(AddressOfBorrowOpaque, "addressOfBorrowOpaque", "", Speci
...
 
+/// createTask<T>(flags: Int,
+///               taskGroup: Builtin.RawPointer? = nil,
+///               initialTaskExecutor: (Builtin.Executor)? = nil,
+///               initialExecutor: (Builtin.Executor)? = nil,
+///               operation: @escaping () async throws -> T)
+///  -> Builtin.NativeObject, Builtin.RawPointer)
+///
+/// Create a new task.
+BUILTIN_SIL_OPERATION(CreateTask, "createTask", Special)
```
Comment (ktoso) [inline include/swift/AST/Builtins.def:963]: hooray for the (both) initial executors <3

File: lib/AST/Builtins.cpp [hunk 2]
```diff
@@ -1510,6 +1510,32 @@ Type swift::getAsyncTaskAndContextType(ASTContext &ctx) {
   return TupleType::get(resultTupleElements, ctx);
 }
 
+static ValueDecl *getCreateTask(ASTContext &ctx, Identifier id) {
+  return getBuiltinFunction(
+      ctx, id, _thin, _generics(_unrestricted, _conformsToDefaults(0)),
+      _parameters(
+        _label("flags", _swiftInt),
+        _label("taskGroup", _defaulted(_optional(_rawPointer), _nil)),
+        //_label("initialExecutor", _defaulted(_optional(_executor), _nil)),
```
Comment (ktoso) [inline lib/AST/Builtins.cpp:1519]: ah okey, so not yet but soon... ok 👍

File: lib/IRGen/GenConcurrency.cpp [hunk 3]
```diff
@@ -404,3 +405,311 @@ void irgen::emitTaskRunInline(IRGenFunction &IGF, SubstitutionMap subs,
...
+  result->setDoesNotThrow();
+  result->setCallingConv(IGF.IGM.SwiftCC);
+
+  // Cast back to NativeObject/RawPointer.
+  auto newTask = IGF.Builder.CreateExtractValue(result, { 0 });
+  newTask = IGF.Builder.CreateBitCast(newTask, IGF.IGM.RefCountedPtrTy);
+  auto newContext = IGF.Builder.CreateExtractValue(result, { 1 });
+  newContext = IGF.Builder.CreateBitCast(newContext, IGF.IGM.Int8PtrTy);
+  return { newTask, newContext };
+}
```
Comment (ktoso) [inline lib/IRGen/GenConcurrency.cpp:715]: thanks for the refactor/unification, reads great

**Review by ktoso — APPROVED** [review]
LGTM, thanks for the refactor, very nice to unify them all 👍

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test Linux

**Comment by rjmccall** [comment]
@swift-ci Please test Linux

**Pull request merged**